spark/src/main/java/org/apache/iceberg/actions/Actions.java

```diff
@@ -19,25 +19,57 @@
 
 package org.apache.iceberg.actions;
 
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
 import org.apache.iceberg.Table;
 import org.apache.spark.sql.SparkSession;
 
 public class Actions {
+  private static final String IMPL_NAME = "SparkActions";
+
+  private static Class<Actions> implClass;
+  private static Constructor<Actions> implConstructor;
 
   private SparkSession spark;
   private Table table;
 
-  private Actions(SparkSession spark, Table table) {
+  protected Actions(SparkSession spark, Table table) {
     this.spark = spark;
     this.table = table;
   }
 
+  private static Constructor actionConstructor() {
```

Comment:
Using our reflection helpers, this could be simpler:

```java
  try {
    DynConstructors.Ctor<Actions> ctor = DynConstructors.builder()
            .impl(SparkActions.class, SparkSession.class, Table.class)
            .build();
    ctor.newInstance(spark, table);
  } catch (NoSuchMethodException e) {
    throw new UnsupportedOperationException("Cannot find Actions implementation on the classpath");
  }
```

Reply:
Unfortunately DynConstructor wraps up exceptions (no explicit throws) a bit differently so we can't just catch a NSME, let me see if I can do something similar.

Reply:
we need build checked :| which has this code which kinda makes me laugh a little
```java
    @SuppressWarnings("unchecked")
    public <C> Ctor<C> buildChecked() throws NoSuchMethodException {
```

Reply:
Yeah, the API allows you to catch checked exceptions if you want to handle them, or the normal path will wrap them with error messages and unchecked exceptions.